### ISSUE TITLE
[Question] How to use MSAL to obtain access token for Azure DevOps REST API? (with support for both personal MSA and Azure AD)

### ISSUE THREAD

**Issue by dpaulino** [AzureAD/microsoft-authentication-library-for-dotnet#1671]
**Which Version of MSAL are you using ?**
MSAL 4.8.2

**Platform**

UWP

**What authentication flow has the issue?**
* Desktop / Mobile
    * [x] Interactive
    * [ ] Integrated Windows Auth
    * [ ] Username Password
    * [ ] Device code flow (browserless)
* Web App
    * [ ] Authorization code
    * [ ] OBO
* Web API
    * [ ] OBO

Other? - please describe;

**Is this a new or existing app?**

This is a new app or experiment

**Repro**

```csharp
public class MsalAuthService : IAdoTokenService
{
    private const string _clientId = "<myClientId>";
    private readonly IPublicClientApplication _publicClientApp;
    private readonly List<string> _scope = new List<string>()
    {
        "499b84ac-1321-427f-aa17-267ca6975798/user_impersonation" // Ref: https://stackoverflow.com/a/53795967/10953422
    };

    public MsalAuthService()
    {
        _publicClientApp = PublicClientApplicationBuilder.Create(_clientId)
            .WithAuthority(AadAuthorityAudience.AzureAdAndPersonalMicrosoftAccount)
            .Build();
    }

    public async Task<string> GetAccessTokenAsync(string accountId, bool silentOnly = true)
    {
        var accounts = await _publicClientApp.GetAccountsAsync();
        IAccount firstAccount = accounts.FirstOrDefault();
        AuthenticationResult result = null;

        try
        {
            result = await _publicClientApp
                .AcquireTokenSilent(_scope, firstAccount)
                .ExecuteAsync();
        }
        catch (MsalUiRequiredException)
        {
            if (silentOnly)
            {
                return null;
            }

            try
            {
                result = await _publicClientApp
                    .AcquireTokenInteractive(_scope)
                    .ExecuteAsync();
            }
            catch (Exception e)
            {
                Debug.WriteLine(e.Message);
                result = null;
            }
        }
        catch
        {
            result = null;
        }

        if (result != null)
        {
            // TODO update cache
        }

        return result?.AccessToken;
    }
}
```

**Expected behavior**
I am expecting to be able to log in with a personal Microsoft Account in the pop up sign in window.

**Actual behavior**
The pop up window shows up but when I enter my personal Microsoft Account email, the window shows an error saying only work/school accounts can be used.

**Possible Solution**

**Additional context/ Logs / Screenshots**
![image](https://user-images.githubusercontent.com/7649589/75619485-2590e780-5b31-11ea-9ecf-40ec0c631497.png)

**Comment by bgavrilMS** [comment]
@jmprieur - I checked and this repros for me as well. Is this resource refusing MSA tokens even though the client allows them?

**Comment by dpaulino** [comment]
Not quite sure what you mean. Can you rephrase? Azure DevOps is supposed to allow MSA authentication

**Comment by dpaulino** [comment]
Would anyone have ideas here?

**Comment by bgavrilMS** [comment]
@henrik-me @jmprieur - I can repro this, but I don't know the reason the resource is refusing MSA accounts. Do you know who we could contact for this?

**Comment by dpaulino** [comment]
I'm reproducing the same problem in xamarin. So this issue does not seem to be specific to the client platform

**Comment by bgavrilMS** [comment]
Hi @dpaulino - no, it's not a problem from the library, it's a problem with the resource. This is why I marked the issue as 'external'. I couldn't find the people who own this - may I suggest you log a bug on Azure Dev Ops instead?

**Comment by dpaulino** [comment]
Looks like ADO is not set up for MSAL. This is really frustrating for me, but I will close this ticket now.

**Comment by stevie1706** [comment]
Adding this to the protected resource map worked for me.  

  ['https://dev.azure.com',["499b84ac-1321-427f-aa17-267ca6975798/user_impersonation", "499b84ac-1321-427f-aa17-267ca6975798/.default"]]

**Comment by KateTang16** [comment]
Is IAdoTokenService from any library?

**Comment by KateTang16** [comment]
Do we have working sample code for ado? I am struggling on this part.

**Comment by MaxymShtoyko** [comment]
@dpaulino hi, did you manage to find a solution to this problem?

**Comment by dpaulino** [comment]
> @dpaulino hi, did you manage to find a solution to this problem?

Yes, don't use MSAL. If you are building an app, create your own webview to manually authenticate to ADO. I used these docs: https://docs.microsoft.com/en-us/azure/devops/integrate/get-started/authentication/oauth?view=azure-devops

**Comment by MaxymShtoyko** [comment]
@jmprieur looks like I don't have access to the AzureAdMultipleOrgs.AnyOrg property cause there is no namespace that contains appropriate enum.

![image](https://user-images.githubusercontent.com/74073293/123110794-167c2a00-d445-11eb-8f29-cfc904c60e2e.png)

**Comment by MaxymShtoyko** [comment]
@jmprieur are there other ways to support a personal Microsoft account email for the "499b84ac-1321-427f-aa17-267ca6975798 / user_impersonation" scope (other than creating your own webview to manually authenticate to ADO).

**Comment by jmprieur** [comment]
Sorry @MaxymShtoyko , I mean `AadAuthorityAudience.AzureAdMultipleOrgs`

**Comment by MaxymShtoyko** [comment]
@jmprieur AadAuthorityAudience.AzureAdMultipleOrgs works, but not in the case of a personal MSA
![image](https://user-images.githubusercontent.com/74073293/123126784-5a296080-d452-11eb-9c86-903bf1141645.png)

**Comment by MaxymShtoyko** [comment]
@jmprieur do you have any other ideas on how to resolve the problem?

**Comment by jmprieur** [comment]
@MaxymShtoyko, given this is a desktop app, I would just use the Azure devops SDKs. Here is a snippet to connect to Azure DevOps interactively and get access to the work items:

```CSharp
        /// <summary>
        /// URL of the Azure DevOps project collection
        /// </summary>
        public string ProjectCollectionUri { get; set; }

        /// <summary>
        /// Project to which to migrate (MUST use the Scrum_Custom process template)
        /// </summary>
        public string Project { get; set; }

        /// <summary>
        /// Connection to Azure DevOps
        /// </summary>
        VssConnection connection;

        /// <summary>
        /// WorkItem service
        /// </summary>
        WorkItemTrackingHttpClient witClient;

        /// <summary>
        /// Connects to Azure DevOps interactively (interactive sign-in)
        /// </summary>
        public void Connect()
        {
            if (connection == null)
            {   
              connection = new VssConnection(new Uri(ProjectCollectionUri), new VssClientCredentials(true));
            }

            // Update the area paths for the project
            witClient = connection.GetClient<WorkItemTrackingHttpClient>();
        }
```

The project should reference the following:

```XML
  <ItemGroup>
    <PackageReference Include="Microsoft.TeamFoundationServer.Client" Version="16.153.0" />
    <PackageReference Include="Microsoft.VisualStudio.Services.Client" Version="16.153.0" />
    <PackageReference Include="Microsoft.VisualStudio.Services.InteractiveClient" Version="16.153.0" />
  </ItemGroup>
```

**Comment by MaxymShtoyko** [comment]
@jmprieur actually I'm trying to integrate MSAL into a mobile application, maybe there are some other ways to help fix this problem in a mobile application besides using Azure DevOps SDKs

**Comment by MaxymShtoyko** [comment]
I noticed that if I leave the scope field blank, the MSAL allows me to use a personal account, but in that case, I'm able to get an appropriate permissions

**Comment by MaxymShtoyko** [comment]
so looks like the problem is related to "499b84ac-1321-427f-aa17-267ca6975798/user_impersonation" scope

**Comment by jmprieur** [comment]
@MaxymShtoyko : the Azure portal seems to recommend: `https://app.vssps.visualstudio.com/user_impersonation`. Did you try this one?
![image](https://user-images.githubusercontent.com/13203188/123231721-c0f15d00-d4d8-11eb-8752-96db25471d3f.png)

**Comment by SheaSmith** [comment]
@jmprieur I tried that scope and it also seems to not allow people with personal accounts to sign in.

**Comment by jmprieur** [comment]
@maliksahil would you know?

**Comment by maliksahil** [comment]
You'll have to use msa-pt I think for MSA accounts.

**Comment by bgavrilMS** [comment]
I think so as well, Azure DevOps does not support MSA logins. There is a workaround for 1st party apps (MSA-passthrough), so this is why Visual Studio is able to connect, but that mechanism is not made available to 3rd parties. 

@SheaSmith - please contact the resource owner Azure DevOps directly ask them to support MSA (personal) accounts. The idenitity SDK does not control this.

**Comment by hickford** [comment]
> I think so as well, Azure DevOps does not support MSA logins. There is a workaround for 1st party apps (MSA-passthrough), so this is why Visual Studio is able to connect, but that mechanism is not made available to 3rd parties.

@bgavrilMS is this still the case? I tried 

![image](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/assets/105314/b1497cb1-9301-439a-a9fa-57d2b3375268)

```go
	client, err := public.New(
		// https://portal.azure.com/#view/Microsoft_AAD_RegisteredApps/ApplicationMenuBlade/~/Overview/appId/81afe8d8-661a-4d77-9769-d541067844b7/isMSAApp~/false
		"81afe8d8-661a-4d77-9769-d541067844b7",
		public.WithAuthority("https://login.microsoftonline.com/consumers/"))
	if err != nil {
		return public.AuthResult{}, err
	}
	// https://learn.microsoft.com/en-us/azure/devops/organizations/accounts/manage-personal-access-tokens-via-api
	scopes := []string{"499b84ac-1321-427f-aa17-267ca6975798/.default"}
	return client.AcquireTokenInteractive(context.Background(), scopes)
```

but got error

> The provided value for the input parameter 'scope' is not valid. The scope '499b84ac-1321-427f-aa17-267ca6975798/.default openid offline_access profile' does not exist.

**Comment by jmprieur** [comment]
@hickford: did you try the .NET SDK for Azure DevOps: https://learn.microsoft.com/azure/devops/integrate/concepts/dotnet-client-libraries?view=azure-devops
